kernel/src/log_segment.rs

```diff
@@ -112,13 +121,234 @@ impl LogSegment {
+    pub(crate) fn with_checkpoint(mut self, checkpoint: CheckpointMetadata) -> Self {
+        self.checkpoint = Some(checkpoint);
+        self
+    }
+
+    /// Optionally set the start version of the [`LogSegment`]. This ensures that all commit files
+    /// are above this version.
+    #[allow(unused)]
+    pub(crate) fn with_start_version(mut self, version: Version) -> Self {
+        self.start_version = Some(version);
+        self
+    }
+    /// Optionally set the end version of the [`LogSegment`]. This ensures that all commit files
+    /// and checkpoints are below the end version.
+    pub(crate) fn with_end_version(mut self, version: Version) -> Self {
+        self.end_version = Some(version);
+        self
+    }
+    /// Optionally specify that the [`LogSegment`] will not have any checkpoint files. It will only
+    /// be made up of commit files.
+    #[allow(unused)]
+    pub(crate) fn omit_checkpoint_files(mut self) -> Self {
+        self.omit_checkpoint_files = true;
+        self
+    }
+
+    /// Optionally specify that the commits in the [`LogSegment`] will be in order. By default, the
+    /// [`LogSegment`] will reverse the order of commit files, with the latest commit coming first.
+    #[allow(unused)]
+    pub(crate) fn with_reversed_commit_files(mut self) -> Self {
+        self.reversed_commit_files = true;
+        self
+    }
+    /// Build the [`LogSegment`]
+    ///
+    /// This fetches checkpoint and commit files using the `fs_client`.
+    pub(crate) fn build(self) -> DeltaResult<LogSegment> {
+        let Self {
+            fs_client,
+            table_root,
+            checkpoint,
+            start_version,
+            end_version,
+            omit_checkpoint_files,
+            reversed_commit_files,
+        } = self;
+        let log_root = table_root.join("_delta_log/").unwrap();
+        let (mut commit_files, mut checkpoint_files) = match (checkpoint, end_version) {
+            (Some(cp), None) => Self::list_log_files_with_checkpoint(&cp, fs_client, &log_root)?,
+            (Some(cp), Some(version)) if cp.version >= version => {
+                Self::list_log_files_with_checkpoint(&cp, fs_client, &log_root)?
+            }
+            _ => {
+                let (commit_files, checkpoint_files, _) =
+                    Self::list_log_files_from_version(fs_client, &log_root, None)?;
+
+                (commit_files, checkpoint_files)
+            }
+        };
+
+        if omit_checkpoint_files {
+            checkpoint_files.clear();
+        }
+
+        // Commit file versions must satisfy the following:
+        // - Be greater than the start version
+        // - Be greater than the most recent checkpoint version if it exists
+        // - Be less than or equal to the end version.
+        if let Some(start_version) = start_version {
+            commit_files.retain(|log_path| start_version <= log_path.version);
```

Comment:
mega nit but this feels more understandable to me
```suggestion
            commit_files.retain(|log_path| log_path.version >= start_version);
```